File: backend/app/schemas/recipe.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Self

from pydantic import BaseModel, Field, model_validator

from .recipe_category import RecipeCategoryResponse
from .recipe_tag import RecipeTagResponse


class RecipeSource(str, Enum):
    cookidoo = "cookidoo"
    manual = "manual"
    web = "web"


class Difficulty(str, Enum):
    easy = "easy"
    medium = "medium"
    hard = "hard"


class IngredientCategory(str, Enum):
    kuehlregal = "kuehlregal"
    obst_gemuese = "obst_gemuese"
    trockenware = "trockenware"
    drogerie = "drogerie"
    sonstiges = "sonstiges"


class IngredientCreate(BaseModel):
    name: str
    amount: float | None = None
    unit: str | None = None
    category: IngredientCategory = IngredientCategory.sonstiges


class IngredientResponse(BaseModel):
    id: int
    name: str
    amount: float | None
    unit: str | None
    category: str

    model_config = {"from_attributes": True}


class RecipeCreate(BaseModel):
    title: str
    source: RecipeSource = RecipeSource.manual
    cookidoo_id: str | None = None
    servings: int = 4
    prep_time_active_minutes: int | None = None
    prep_time_passive_minutes: int | None = None
    difficulty: Difficulty = Difficulty.medium
    instructions: str | None = None
    notes: str | None = None
    image_url: str | None = None
    ai_accessible: bool = True
    recipe_category_id: int | None = None
    tag_ids: list[int] = []
    ingredients: list[IngredientCreate] = []


class RecipeUpdate(BaseModel):
    title: str | None = None
    servings: int | None = None
    prep_time_active_minutes: int | None = None
    prep_time_passive_minutes: int | None = None
    difficulty: Difficulty | None = None
    instructions: str | None = None
    notes: str | None = None
    image_url: str | None = None
    ai_accessible: bool | None = None
    recipe_category_id: int | None = None
    tag_ids: list[int] | None = None
    ingredients: list[IngredientCreate] | None = None


class RecipeResponse(BaseModel):
    id: int
    title: str
    source: str
    cookidoo_id: str | None
    servings: int
    prep_time_active_minutes: int | None
    prep_time_passive_minutes: int | None
    difficulty: str
    last_cooked_at: datetime | None
    cook_count: int
    instructions: str | None
    notes: str | None
    image_url: str | None
    cover_image_path: str | None = Field(default=None, exclude=True)
    ai_accessible: bool
    recipe_category_id: int | None
    category: RecipeCategoryResponse | None
    tags: list[RecipeTagResponse]
    ingredients: list[IngredientResponse]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

    @model_validator(mode="after")
    def _apply_cover_image_url(self) -> Self:
        if self.cover_image_path:
            return self.model_copy(update={"image_url": f"/api/recipes/{self.id}/cover"})
        return self


class CookingHistoryResponse(BaseModel):
    id: int
    recipe_id: int
    cooked_at: datetime
    servings_cooked: int
    rating: int | None
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class RecipeDetailResponse(RecipeResponse):
    history: list[CookingHistoryResponse] = []


class RecipeSuggestion(BaseModel):
    id: int
    title: str
    difficulty: str
    prep_time_active_minutes: int | None
    image_url: str | None = None
    last_cooked_at: datetime | None
    cook_count: int
    days_since_cooked: int | None

    model_config = {"from_attributes": True}


class UrlImportRequest(BaseModel):
    url: str


class UrlImportPreview(BaseModel):
    title: str
    servings: int = 4
    prep_time_active_minutes: int | None = None
    prep_time_passive_minutes: int | None = None
    difficulty: str = "medium"
    instructions: str | None = None
    image_url: str | None = None
    source_url: str | None = None
    ingredients: list[IngredientCreate] = []

File: obsolete/ios/Familienkalender/Models/Recipe.swift
import Foundation

enum RecipeSource: String, Codable {
    case manual, cookidoo, web

    var displayName: String {
        switch self {
        case .manual: "Manuell"
        case .cookidoo: "Cookidoo"
        case .web: "Web"
        }
    }
}

enum Difficulty: String, Codable, CaseIterable {
    case easy, medium, hard

    var displayName: String {
        switch self {
        case .easy: "Einfach"
        case .medium: "Mittel"
        case .hard: "Aufwendig"
        }
    }

    var color: String {
        switch self {
        case .easy: "#00875A"
        case .medium: "#FF8B00"
        case .hard: "#DE350B"
        }
    }
}

enum IngredientCategory: String, Codable, CaseIterable {
    case kuehlregal
    case obstGemuese
    case trockenware
    case drogerie
    case sonstiges

    var displayName: String {
        switch self {
        case .kuehlregal: "Kuehlregal"
        case .obstGemuese: "Obst & Gemuese"
        case .trockenware: "Trockenware"
        case .drogerie: "Drogerie"
        case .sonstiges: "Sonstiges"
        }
    }

    var icon: String {
        switch self {
        case .kuehlregal: "🧊"
        case .obstGemuese: "🍇"
        case .trockenware: "🍞"
        case .drogerie: "🧴"
        case .sonstiges: "📦"
        }
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let raw = try container.decode(String.self)
        switch raw {
        case "kuehlregal": self = .kuehlregal
        case "obst_gemuese": self = .obstGemuese
        case "trockenware": self = .trockenware
        case "drogerie": self = .drogerie
        case "sonstiges": self = .sonstiges
        default: self = .sonstiges
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .kuehlregal: try container.encode("kuehlregal")
        case .obstGemuese: try container.encode("obst_gemuese")
        case .trockenware: try container.encode("trockenware")
        case .drogerie: try container.encode("drogerie")
        case .sonstiges: try container.encode("sonstiges")
        }
    }

    init?(apiValue: String) {
        switch apiValue {
        case "kuehlregal": self = .kuehlregal
        case "obst_gemuese": self = .obstGemuese
        case "trockenware": self = .trockenware
        case "drogerie": self = .drogerie
        case "sonstiges": self = .sonstiges
        default: return nil
        }
    }

    var apiValue: String {
        switch self {
        case .kuehlregal: return "kuehlregal"
        case .obstGemuese: return "obst_gemuese"
        case .trockenware: return "trockenware"
        case .drogerie: return "drogerie"
        case .sonstiges: return "sonstiges"
        }
    }
}

struct IngredientCreate: Codable {
    let name: String
    let amount: Double?
    let unit: String?
    var category: String = "sonstiges"
}

struct IngredientResponse: Codable, Identifiable {
    let id: Int
    let name: String
    let amount: Double?
    let unit: String?
    let category: String

    var categoryEnum: IngredientCategory {
        IngredientCategory(apiValue: category) ?? .sonstiges
    }
}

struct RecipeCreate: Codable {
    let title: String
    var source: String = "manual"
    var cookidooId: String?
    var servings: Int = 4
    var prepTimeActiveMinutes: Int?
    var prepTimePassiveMinutes: Int?
    var difficulty: String = "medium"
    var instructions: String?
    var notes: String?
    var imageUrl: String?
    var aiAccessible: Bool = true
    var ingredients: [IngredientCreate] = []
}

struct RecipeUpdate: Codable {
    var title: String?
    var servings: Int?
    var prepTimeActiveMinutes: Int?
    var prepTimePassiveMinutes: Int?
    var difficulty: String?
    var instructions: String?
    var notes: String?
    var imageUrl: String?
    var aiAccessible: Bool?
    var ingredients: [IngredientCreate]?
}

struct RecipeResponse: Codable, Identifiable {
    let id: Int
    let title: String
    let source: String
    let cookidooId: String?
    let servings: Int
    let prepTimeActiveMinutes: Int?
    let prepTimePassiveMinutes: Int?
    let difficulty: String
    let lastCookedAt: String?
    let cookCount: Int
    let instructions: String?
    let notes: String?
    let imageUrl: String?
    let aiAccessible: Bool
    let ingredients: [IngredientResponse]
    let createdAt: String
    let updatedAt: String

    var difficultyEnum: Difficulty {
        Difficulty(rawValue: difficulty) ?? .medium
    }

    var sourceEnum: RecipeSource {
        RecipeSource(rawValue: source) ?? .manual
    }

    var formattedPrepTime: String? {
        let active = prepTimeActiveMinutes ?? 0
        let passive = prepTimePassiveMinutes ?? 0
        if active == 0 && passive == 0 { return nil }
        if passive == 0 { return "\(active) Min." }
        return "\(active) Min. aktiv, \(passive) Min. passiv"
    }
}

struct RecipeSuggestion: Codable, Identifiable {
    let id: Int
    let title: String
    let difficulty: String
    let prepTimeActiveMinutes: Int?
    let lastCookedAt: String?
    let cookCount: Int
    let daysSinceCooked: Int?
}

struct UrlParseRequest: Codable {
    let url: String
}

struct UrlImportPreview: Codable {
    let title: String?
    let servings: Int?
    let prepTimeActiveMinutes: Int?
    let prepTimePassiveMinutes: Int?
    let difficulty: String?
    let instructions: String?
    let imageUrl: String?
    let ingredients: [IngredientCreate]?
}
